text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+data processing, data minimization). All existing components will continue to be compatible and operational.
+
+A two-phase integration is proposed to allow incremental benefits.
+
+#### Phase 1
+
+This proposal is designed as a protocol extension compatible with the existing OTLP protocol. As illustrated in the
+following diagram, a new OTLP-Arrow to OTLP receiver will be responsible for translating the protocol extension to the
+existing protocol. Similarly, a new exporter will be responsible for translating the OTLP messages into this new Arrow-based
+format.
+
+![OTEL Collector](img/0156_OTEL%20-%20OTEL%20Collector.png)
+
+This first step is intended to address the specific use cases of **traffic reduction** and native support of
+**multivariate time-series**. Based on community feedback, many companies want to reduce the cost of transferring
+telemetry data over the Internet. By adding a collector that acts as a point of integration and traffic conversion at
+the edge of a client environment, we can take advantage of the columnar format to eliminate redundant data and optimize
+the compression ratio. This is illustrated in the following diagram.
+
+![Traffic reduction](img/0156_OTEL%20-%20InternetTrafficReduction.png)
+
+> Note 1: A fallback mechanism can be used to handle the case where the new protocol is not supported by the target.
+> More on this mechanism in this [section](#protocol-extension-and-fallback-mechanism).
+
+#### Phase 2
+
+Phase 2 aims to extend the support of Apache Arrow end-to-end and more specifically inside the collector to better
+support the following scenarios: cost efficiency, advanced data processing, data minimization. New receivers, processors,
+and exporters supporting Apache Arrow natively will be developed. A bidirectional adaptation layer OTLP / OTLP Arrow
+will be developed within the collector to continue supporting the existing ecosystem. The following diagram is an
+overview of a collector supporting both OTLP and an end-to-end OTLP Arrow pipeline.
+
+![OTEL Arrow Collector](img/0156_OTEL%20-%20OTELArrowCollector.png)
+
+Implementing an end-to-end column-oriented pipeline will provide many benefits such as:
+
+- **Accelerate stream processing**,
+- **Reduce CPU and Memory usage**,
+- **Improve compression ratio end-to-end**,
+- **Access to the Apache Arrow ecosystem** (query engine, parquet support, ...).
+
+## Protocol Details
+
+The protocol specifications are composed of two parts. The first section describes the new gRPC service supporting
+column-oriented telemetry data. The second section presents the mapping between the OTLP entities and their Apache
+Arrow counterpart.
+
+### ArrowStreamService
+
+OTLP Arrow defines the columnar encoding of telemetry data and the gRPC-based protocol used to exchange data between
+the client and the server. OTLP Arrow is a bi-directional stream oriented protocol leveraging Apache Arrow for the
+encoding of the telemetry data.
```

Comment:
Arrow Flight RPC was considered at the beginning of the project but a tighter integration with the existing OTLP/RPC protocol was preferred for the following reasons:
* The OTEL collector is written in Go but the client SDKs are supported for about 15 languages. This was far from being the case for Arrow Flight (1 year ago) and it is still the case (5 languages supported in early 2023 and among those C# does not support the DoExchange method you mentioned). 
* The Arrow Flight protocol is also much more elaborate and supports more usage scenarios, in any case more complex than what we really need for OTEL. This may seem like an advantage, but it also means a potentially greater fragility and a larger attack surface.
* The fallback mechanism between the two protocols is also an aspect that I found more complex to implement with Arrow Flight RPC.  

In retrospect, I still think the current choice is the right one. It gives us more control over what we need to support.